lib/spack/docs/tutorial_modules.rst

```diff
@@ -1,52 +1,277 @@
+module files are generated by Spack. The following diagram provides a high-level view
+of the process:
+
+
+.. image:: module_file_generation.*
+
+The red dashed line above represents Spack's boundaries, the blue one Spack's vendored
+dependencies. In short, module files are generated combining:
+
+  * the configuration details in ``config.yaml`` and ``modules.yaml``
+  * the information contained in Spack packages (and processed by the module subpackage)
+  * a set of `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_ templates
+
+with the vendored template engine to "stamp out" each particular module file. As Spack serves
+very diverse needs this process has many points of customization, and we'll explore most of
+them in the next sections.
 
 .. _module_file_tutorial_prerequisites:
 
--------------
-Prerequisites
--------------
+-----------------------
+Set up for the tutorial
+-----------------------
+
+In order to showcase the capabilities of Spack's module file generation, we need
+a representative set of software to work with. This set must include different
+flavors of the same packages installed alongside each other and some
+:ref:`external packages <sec-external-packages>`.
 
-Before proceeding further ensure:
+The purpose of this setup is not to make our life harder but to demonstrate
+how Spack can help with similar situations, as they will happen on real HPC clusters.
+For instance, it's very common the case in which the MPI implementation
+is already provided by vendors and is better not to be built again by Spack.
 
-- you have LMod or Environment Modules available
-- have :ref:`shell support <shell-support>` activated in Spack
+The best way to follow along is to use a Docker image, which comes
+with Spack and all the software used in the following parts already
+pre-installed. If you want to proceed this way, read :ref:`module_file_tutorial_use_docker`.
 
-If you need to install Lmod or Environment module you can refer
-to the documentation :ref:`here <InstallEnvironmentModules>`.
+If you don't have Docker installed or for any other reason you
+prefer to work locally, follow instead :ref:`module_file_tutorial_work_locally`
+to know how to clone Spack and install the software. Be aware that in this case
+the set-up will take sensibly longer. Furthermore, all the snippets afterwards
+come from the Docker image, so your progress may vary.
 
+.. _module_file_tutorial_use_docker:
 
 ^^^^^^^^^^^^^^^^^^
-Add a new compiler
+Use a Docker image
 ^^^^^^^^^^^^^^^^^^
 
-Spack automatically scans the environment to search for available
-compilers on first use. On Ubuntu 14.04, a fresh clone will show
-something like this:
+The fastest way to set-up your environment is to :ref:`use a Docker image <workflow_create_docker_image>`:
+
+.. code-block:: console
+
+  $ docker pull alalazo/spack:module_tutorial
+  $ docker run -h module-file-tutorial -it alalazo/spack:module_tutorial
```

Comment:
Perhaps add `--rm` to this so that the container is automagically cleaned up when it exits?